### PR TITLE
fix delegator prisma structure

### PR DESCRIPTION
**Pull request by bayram98**
# Description

This PR is for to add `whitelist` `m-m` implicit relationship between `Contract` and `Reporter` tables.

```mermaid
graph TD;
	organization-->reporter
	organization-->reporter1
        reporter-->whitelist
        whitelist-->contract
        reporter1-->whitelist
        function-->transaction
        contract-->transaction
	contract-->function
	organization1 --> reporter2
        whitelist --> contract1
        reporter2 --> whitelist
        contract1-->function1
        contract1-->transaction1
        function1-->transaction1
```

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist before requesting a review

- [x] I have performed a self-review of my code.

### CODE REVIEW

File: delegator/prisma/schema.prisma [hunk 1]
```diff
@@ -58,19 +44,33 @@ model Function {
...
+  s           String    @db.VarChar(66)
+  rawTx       String    @db.VarChar(1024)
+  signedRawTx String?   @db.VarChar(1024)
+  succeed     Boolean?
+  function    Function? @relation(fields: [functionId], references: [id])
+  functionId  BigInt?   @map("function_id")
+  contract    Contract? @relation(fields: [contractId], references: [id])
+  contractId  BigInt?   @map("contract_id")
+  reporter    Reporter? @relation(fields: [reporterId], references: [id])
+  reporterId  BigInt?   @map("reporterId")
```
Comment (martinkersner) [inline delegator/prisma/schema.prisma:73]: Please unify the formatting. `function_id` and `contract_id` have underscore.

```suggestion
  reporterId  BigInt?   @map("reporter_id")
```

File: delegator/prisma/schema.prisma [hunk 2]
```diff
@@ -7,47 +7,33 @@ datasource db {
...
-  address     String        @unique @db.VarChar(42)
-  Function    Function[]
-  Reporter    Reporter[]
-  Transaction Transaction[]
+  id                BigInt        @id @default(autoincrement()) @map("contract_id")
+  address           String        @unique @db.VarChar(42)
+  allowAllFunctions Boolean?      @default(false)
+  function          Function[]
+  transaction       Transaction[]
+  reporter          Reporter[]    @relation(name: "WhitelistTable")
```
Comment (martinkersner) [inline delegator/prisma/schema.prisma:36]: I think we do not need to name it table, and we should avoid capital letters. We could name it maybe `contract_reporter_whitelist`. What do you think?

**Review by martinkersner — CHANGES_REQUESTED** [review]

File: delegator/src/sign/sign.service.ts [hunk 3]
```diff
@@ -118,22 +102,18 @@ export class SignService {
...
+    const contract = await this.prisma.contract.findUnique({
+      where: { address: tx.to },
+      include: {
+        reporter: { where: { address: tx.from } },
+        function: { where: { encodedName } }
       }
     })
-    if (result.length == 0) {
+
+    if (contract && contract.reporter.length != 0 && contract.function.length != 0) {
```
Comment (martinkersner) [inline delegator/src/sign/sign.service.ts:114]: I think we do not need to test on `contract.reporter.length != 0 && contract.function.length != 0`

File: delegator/src/sign/sign.service.ts [hunk 4]
```diff
@@ -21,14 +21,15 @@ export class SignService {
 
   async create(data: SignDto) {
     const transaction = await this.prisma.transaction.create({ data })
+    let validatedQuery
```
Comment (martinkersner) [inline delegator/src/sign/sign.service.ts:24]: I think the whole function should be inside of `try-catch` block. Each function could theoretically raise an error, we should be able to catch all. Also, it makes it easier to read.

```typescript
async create(data: SignDto) {
  try {
    const transaction = await this.prisma.transaction.create({ data })
    const validatedQuery = await this.validateTransaction(transaction)
    const signedRawTx = await this.signTxByFeePayer(transaction)
    return await this.updateTransaction(transaction, signedRawTx, validatedQuery)
  } catch (e) {
    const msg = `DelegatorError: ${e.name}`
    throw new HttpException(msg, HttpStatus.BAD_REQUEST)
  }
}
```

File: delegator/src/contract/dto/contract.dto.ts [hunk 5]
```diff
@@ -3,4 +3,7 @@ import { ApiProperty } from '@nestjs/swagger'
 export class ContractDto {
   @ApiProperty()
   address: string
+
+  @ApiProperty()
+  allowAllFunctions: boolean
```
Comment (martinkersner) [inline delegator/src/contract/dto/contract.dto.ts:8]: Do we still use it? I think you removed it from the validation function.

File: delegator/src/contract/contract.service.spec.ts [hunk 6]
```diff
@@ -27,12 +27,18 @@ describe('ContractService', () => {
 
   it('should insert new Contract', async () => {
     const address = '0x0000000000000000000000000000000000000001'
-    const contractData = await contract.create({ address })
+    const contractData = await contract.create({
+      address,
+      allowAllFunctions: false
```
Comment (martinkersner) [inline delegator/src/contract/contract.service.spec.ts:32]: Are we still using the `allowAllFunctions`?

File: delegator/src/contract/contract.service.spec.ts [hunk 7]
```diff
@@ -27,12 +27,18 @@ describe('ContractService', () => {
...
+    })
     expect(contractData.address).toBe(address)
 
     // The same Contract cannot be defined twice
     await expect(async () => {
-      await contract.create({ address })
+      await contract.create({
+        address,
+        allowAllFunctions: false
```
Comment (martinkersner) [inline delegator/src/contract/contract.service.spec.ts:40]: Are we still using the `allowAllFunctions`?

**Review by martinkersner — CHANGES_REQUESTED** [review]

File: delegator/src/contract/dto/contract.dto.ts (hunk 5 above)
Comment (bayram98) [reply delegator/src/contract/dto/contract.dto.ts]: Yes. I decided to remove logic of `allowAllFunctions`. 
I removed all with latest changes.

File: delegator/src/sign/sign.service.ts (hunk 3 above)
Comment (bayram98) [reply delegator/src/sign/sign.service.ts]: The previous approach of how I filter query was not correct. 

I fixed the issue, and here is the correct way of how to filter query from relationships. 

```Typescript
      const relationQuery = await this.prisma.contract.findMany({
      where: {
        address: tx.to,
        reporter: { some: { address: tx.from } },
        function: { some: { encodedName } }
      },
      include: {
        reporter: { where: { address: tx.from } },
        function: { where: { encodedName } }
      }
    })
```

Comment (martinkersner) [reply delegator/src/sign/sign.service.ts]: Great! ❤️

**Review by martinkersner — APPROVED** [review]
LGTM!

**Pull request merged**